### PR TITLE
Refine the logic of device construction when only device index is given

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129119
* #129363
* #129205
* #129463

# Motivation
Before this PR, device construction was `cuda` type when only a device index was given. It also returns the `PrivateUser1` type if a `PrivateUser1` type is registered.
```bash
>>> import torch
>>> device = torch.device(0)
>>> device.type
'cuda'
>>> a = torch.tensor([1, 2])
>>> b = a.to(0)
>>> b
tensor([1, 2], device='cuda:0')
```
It works well on CUDA GPU. But it will raise unexpected information and error running on XPU.
```bash
>>> import torch
>>> device = torch.device(0)
>>> device.type
'cuda'
>>> a = torch.tensor([1, 2])
>>> b = a.to(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/xxx/pytorch/torch/cuda/__init__.py", line 302, in _lazy_init
    raise AssertionError("Torch not compiled with CUDA enabled")
AssertionError: Torch not compiled with CUDA enabled
```
With this PR, refine the logic to use the currently available device type instead.

### CODE REVIEW

File: torch/csrc/utils/python_arg_parser.h [hunk 1]
```diff
@@ -813,8 +813,14 @@ inline at::Device toDevice(PyObject* obj) {
           c10::DeviceType::PrivateUse1,
           static_cast<c10::DeviceIndex>(device_index));
     }
+#ifdef USE_CUDA
```
Comment (albanD) [inline torch/csrc/utils/python_arg_parser.h:816]: Updated https://github.com/pytorch/pytorch/issues/126646 with details.
Can you change this to get the current accelerator instead?

Comment (guangyey) [reply torch/csrc/utils/python_arg_parser.h]: OK, using the current accelerator looks better.
And I add `XPU` to `getAccelerator` and refine the code.

File: docs/source/tensor_attributes.rst [hunk 2]
```diff
@@ -213,7 +213,8 @@ non-None device argument.  To globally change the default device, see also
 
 .. note::
    For legacy reasons, a device can be constructed via a single device ordinal, which is treated
-   as a cuda device.  This matches :meth:`Tensor.get_device`, which returns an ordinal for cuda
+   as a currently available device type (i.e. "cuda" if cuda is available, "xpu" if xpu is available).
```
Comment (albanD) [inline docs/source/tensor_attributes.rst:216]: Given that we're going to have a few features relying on it as the default (pin memory for example), I think we should make "accelerator" a public concept in this doc.
I think we can:
- Have one paragraph in https://pytorch.org/docs/stable/torch.html that introduces the concept of accelerator and lists all the current ones (based on the list in c10).
- Have other docs like this one just mention the "current accelerator" with a link to the paragraph above.

WDYT?

Comment (gujinghui) [reply docs/source/tensor_attributes.rst:216]: Sounds good. Maybe we can take below RFC together with this PR, to figure out an accelerator mechanism both in frontend and backend? https://github.com/pytorch/pytorch/issues/128403

@guangyey let's have a talk next week.

Comment (guangyey) [reply docs/source/tensor_attributes.rst:216]: It is great to me~

File: docs/source/tensor_attributes.rst [hunk 3]
```diff
@@ -213,7 +213,8 @@ non-None device argument.  To globally change the default device, see also
 
 .. note::
    For legacy reasons, a device can be constructed via a single device ordinal, which is treated
-   as a cuda device.  This matches :meth:`Tensor.get_device`, which returns an ordinal for cuda
+   as a default :ref:`accelerator<accelerators>` type.
```
Comment (albanD) [inline docs/source/tensor_attributes.rst:216]:
```suggestion
   as the current :ref:`accelerator<accelerators>` type.
```

**Review by albanD — APPROVED** [review]
So much simpler! 
Thanks!

**Review by gujinghui — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]

**Comment by dvrogozh** [comment]
Folks, just want to give you a heads up that there was an assumption that this PR will address https://github.com/huggingface/transformers/issues/31941. Unfortunately, this PR does not fully solve it. I believe that PR gives an essential change, but it alone is not enough to get the described case working. This being said, mind that at the moment I don't know what's the remainder of the root cause for 31941. It might be that HF will need another fix somewhere or pytorch - can't say right now.

**Comment by dvrogozh** [comment]
> Folks, just want to give you a heads up that there was an assumption that this PR will address https://github.com/huggingface/transformers/issues/31941. Unfortunately, this PR does not fully solve it.

Update on this. The root cause is the bug in HF safetensors library, see https://github.com/huggingface/safetensors/issues/499. This PR indeed provides required prerequisite for the fix. So, thank you for merging @gujinghui.

**Pull request closed without merging**